share/poly.go

```diff
@@ -125,19 +129,35 @@ func (p *PriPoly) Commit(b abstract.Point) *PubPoly {
 	return &PubPoly{p.g, b, commits}
 }
 
+// Mul multiples p  and p2 together. The result is a polynomial of the sum of
+// the two degrees of p and p2. NOTE: it does not check for null coefficients
+// after the multiplication, so the degree of the polynomial is "always" as
+// described above. This is only to use in secret sharing schemes, and is not to
+// be considered a general polynomial manipulation routine.
+func (p *PriPoly) Mul(q *PriPoly) *PriPoly {
+	d1 := len(p.coeffs) - 1
+	d2 := len(q.coeffs) - 1
+	newDegree := d1 + d2
+	coeffs := make([]abstract.Scalar, newDegree+1)
+	for i := range coeffs {
+		coeffs[i] = p.g.Scalar().Zero()
+	}
+	for i := range p.coeffs {
+		for j := range q.coeffs {
+			tmp := p.g.Scalar().Mul(p.coeffs[i], q.coeffs[j])
+			coeffs[i+j] = tmp.Add(coeffs[i+j], tmp)
+		}
+	}
+	return &PriPoly{p.g, coeffs}
+}
+
 // RecoverSecret reconstructs the shared secret p(0) from a list of private
 // shares using Lagrange interpolation.
 func RecoverSecret(g abstract.Group, shares []*PriShare, t, n int) (abstract.Scalar, error) {
-	x := make(map[int]abstract.Scalar)
-	for i, s := range shares {
-		if s == nil || s.V == nil || s.I < 0 || n <= s.I {
-			continue
-		}
-		x[i] = g.Scalar().SetInt64(1 + int64(s.I))
-	}
+	x := xScalar(g, shares, t, n)
 
-	if len(x) < t {
-		return nil, errors.New("not enough good private shares to reconstruct shared secret")
+	if len(x) != t {
```

Comment:
I talked too fast. In fact, there is a call to `xScalar` before which only takes up to `t` shares, so that's why we compare for equality instead of doing `if len(x) <= t`. 
In short, it's fine :)